calcom/cal.com
# feat: Message Queuing System using Trigger.dev

## What does this PR do?

Add Trigger.dev as a queue system.

### Why trigger.dev?

First of all, they are Open Source!! 🎉🎉🎉

Second, they are a great fit for our needs. We need a queue system that can handle a lot of jobs and that can scale easily. We also need a queue system that can handle webhooks and that can be integrated with other services. Trigger.dev can do all of that and more.

Video:
[![Trigger.dev](https://img.youtube.com/vi/cT1pfsaZtdA/0.jpg)](https://www.youtube.com/watch?v=cT1pfsaZtdA)



### What else needs to be done:
- [ ] Configure adicional integrations like Twilio, calendars and so on.
- [ ] Design triggers accordantly to each job to allow individual logging / debugging / retry / testing. For example, emails are currently being build when the `sendEmail` is called so pretty much all emails would be going through the same job "send email" and making it hard to identify stuff.
- [ ] Change current flows to appropriately add itens to queue (trigger) when Trigger.dev configurations are set.

Fixes # (issue)
<https://github.com/calcom/cal.com/issues/12556>


## Requirement/Documentation

<https://github.com/calcom/cal.com/issues/12556>



## Type of change




## How should this be tested?



Start the local version of Trigger.dev or [create an account](https://trigger.dev) on the cloud version.:

```bash
cd /infra/docker/queue
docker-compose up -d
```

The following environment variables are required to run this can be found in the `Environments & API Keys` section of the dashboard:

- `TRIGGER_API_KEY`: key of the environment/project in Trigger.dev.
- `NEXT_PUBLIC_TRIGGER_PUBLIC_API_KEY`: Public key of the environment/project in Trigger.dev (used in the frontend).
- `TRIGGER_API_URL`: URL of the Trigger.dev API. Defaults to `https://localhost:3030` if you are using the local version or you can use `https://cloud.trigger.dev`.
- `TRIGGER_PROJECT_ID`: ID of the project in Trigger.dev. Can be found below the Endpoints heading after adding the first endpoint. If you using the local docker, start the web service `yarn dev` and then add the following URL `http://host.docker.internal:3000/api/trigger`. If you are using the cloud version, you might need to run ngrok to expose the local web service to the internet.

After you have added the environment variables, you can restart the web service and see the `example` job created by the file `packages/queue/jobs/examples.ts`.

## Mandatory Tasks


## Checklist



- I haven't checked if my PR needs changes to the documentation
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes
